### PR TITLE
Fix IPC4 capture regression

### LINKED ISSUE #4261: [BUG] I/O error with headset capture stop/start on MTL
**Describe the bug**
This is a driver regression on all MTL platforms. It's caused by https://github.com/thesofproject/linux/commit/658d0f0a1d4e
```
ASoC: SOF: pcm: Make hw_params reset conditional for IPC3
In the case of IPC4, since there is no PCM_PARAMS IPC to send the new
stream tag when restarting a stream without a hw_free, the original
stream tag needs to be preserved. So, add new a flag as part of struct
sof_ipc_pcm_ops, reset_hw_params_during_stop and set it only for IPC3.
This will ensure that the host DMA stream tag will not be given up during
the STOP trigger for IPC4.

```
I/O error occurs when testing signal stop/start on headset capture pipeline at the first round. 

**console log**
```
2023-03-16 11:08:12 UTC [REMOTE_INFO] ktime=1665 sof-test PID=11448: starting
2023-03-16 11:08:15 UTC [REMOTE_INFO] Starting /usr/local/bin/mtrace-reader.py
2023-03-16 11:08:15 UTC [REMOTE_INFO] /home/ubuntu/sof-test/test-case/check-signal-stop-start.sh will use topology /usr/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt711-4ch.tplg to run the test case
2023-03-16 11:08:15 UTC [REMOTE_INFO] Pipeline list to ignore is specified, will ignore 'pcm=HDA Digital' in test case
2023-03-16 11:08:15 UTC [REMOTE_INFO] Run command to get pipeline parameters
2023-03-16 11:08:15 UTC [REMOTE_COMMAND] sof-tplgreader.py /usr/lib/firmware/intel/sof-ace-tplg/sof-mtl-rt711-4ch.tplg -f 'type:capture & ~pcm:HDMI' -b ' pcm:HDA Digital' -s 0 -e
2023-03-16 11:08:15 UTC [REMOTE_INFO] Testing: run stop/start test on PCM:Jack in,capture. Interval time: 0.5
2023-03-16 11:08:15 UTC [REMOTE_COMMAND] arecord -Dhw:0,1 -r 48000 -c 2 -f S16_LE /dev/null -q &
2023-03-16 11:08:16 UTC [REMOTE_INFO] Stop/start count: 1
overrun!!! (at least 125.449 ms long)
arecord: pcm_read:2221: read error: Input/output error
process: 11545, status: not found
2023-03-16 11:08:17 UTC [REMOTE_ERROR] arecord(11545) process is in an abnormal status
```

**mtrace**
```
[00:00:00.379,535] <inf> ipc: ipc_cmd: rx	: 0x13000004|0x1
[00:00:00.379,601] <inf> pipe: pipeline_trigger: pipe:1 0x0 pipe trigger cmd 7
[00:00:00.379,620] <inf> ll_schedule: zephyr_ll_task_schedule_common: task add 0xa0112ec0 0x201b0U priority 0 flags 0x0
[00:00:00.379,668] <inf> ll_schedule: zephyr_domain_register: zephyr_domain_register domain->type 1 domain->clk 3 domain->ticks_per_ms 38400 period 1000
[00:00:00.379,781] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.379,833] <inf> dai_comp: dai_capture_params: comp:0 0x40000 dai_capture_params() src_dev = 23 stream_id = 8 src_width = 4 dest_width = 4
[00:00:00.379,885] <inf> eq_iir: eq_iir_prepare: comp:0 0x100000 eq_iir_prepare(), source_format=2, sink_format=2
[00:00:00.379,903] <inf> eq_iir: eq_iir_init_coef: comp:0 0x100000 eq_iir_init_coef(): 1 responses, 2 channels, stream 2 channels
[00:00:00.379,918] <inf> eq_iir: eq_iir_init_coef: comp:0 0x100000 eq_iir_init_coef(), ch 0 is set to response 0
[00:00:00.379,933] <inf> eq_iir: eq_iir_init_coef: comp:0 0x100000 eq_iir_init_coef(), ch 1 is set to response 0
[00:00:00.379,988] <inf> pipe: pipeline_trigger: pipe:0 0x0 pipe trigger cmd 7
[00:00:00.380,006] <inf> ll_schedule: zephyr_ll_task_schedule_common: task add 0xa0113940 0x201b0U priority 0 flags 0x0
[00:00:00.380,766] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.380,826] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.381,760] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.381,778] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.382,796] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.382,816] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.383,810] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.383,830] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.384,751] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.384,776] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.385,751] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.385,771] <wrn> dai_comp: comp:0 0x40000 dai_copy(): nothing to copy
[00:00:00.386,751] <inf> host_comp: host_get_copy_bytes_normal: comp:1 0x40001 no bytes to copy, available bytes: 0, free_bytes: 768
[00:00:00.901,970] <inf> ipc: ipc_cmd: rx	: 0x13000003|0x1
```

**To Reproduce**
~/sof-test/test-case/check-signal-stop-start.sh -m capture -c 50

**Reproduction Rate**
100%

### PR DESCRIPTION
**Pull request by ranj063**
Fixes https://github.com/thesofproject/linux/issues/4261

### CODE REVIEW

File: sound/soc/sof/ipc4-pcm.c [hunk 1]
```diff
@@ -733,7 +711,8 @@ static int sof_ipc4_pcm_hw_params(struct snd_soc_component *component,
 
 	sof_ipc4_build_time_info(sdev, &spcm->stream[substream->stream]);
 
-	return 0;
+	return sof_ipc4_trigger_pipelines(component, substream, SOF_IPC4_PIPE_PAUSED,
+					  SNDRV_PCM_TRIGGER_START);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c]: This needs a big explanation, imho. Triggering in hw_params?

File: sound/soc/sof/ipc4-pcm.c [hunk 2]
```diff
@@ -302,17 +300,6 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
...
-	pipe_widget = spipe->pipe_widget;
-	pipeline = pipe_widget->private;
-
-	/*
-	 * If use_chain_dma attribute is set we proceed to chained DMA
-	 * trigger function that handles the rest for the substream.
-	 */
-	if (pipeline->use_chain_dma)
-		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);
-
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:315]: It might be better to separate the second half of this function as a new one to let's say set_pipelines_state_worker or something?
It is starting to get odd to see triggers all around the code at callback levels which should not trigger at all.

**Review by ujfalusi — COMMENTED** [review]
@ranj063, if I read between the lines the issue is that with IPC4 we need a sequence of 3 and not two, like with IPC3?
IPC3 start:
trigger host
trigger DSP
IPC4:
trigger DSP
trigger host
trigger DSP

Or the IPC3 hw_params does also DMA configuration via the SOF_IPC_STREAM_PCM_PARAMS message, which is done by the pipeline_state change message (from ununit/stopped to PAUSED perhaps) in IPC4?

If this is the case, then I would make this clean and split the ipc4 trigger pipelines so a new part can be used as set state and in hw_params we would tell it to move to paused (or it if was in paused to stopped then back to paused) as a means to configure the DMA in DSP.

Having code calling trigger from various PCM state levels and contexts will just make things hard to follow and leads to issues and fixes like you are reverting in this series.

**Comment by ujfalusi** [comment]
@ranj063, the test results are not too great, should we have a revert separated and figure out working fix?

File: sound/soc/sof/ipc4-pcm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c]: I was a bit too ambitious trying to avoid a pre/post trigger ops. But now it looks like it might be the best route to go. Updated now

File: sound/soc/sof/pcm.c [hunk 3]
```diff
@@ -336,21 +333,33 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
...
+	case SNDRV_PCM_TRIGGER_START:
+		/* invoke platform trigger to start DMA only if pre_trigger is successful */
+		if (ret < 0)
+			return ret;
 		snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		break;
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_STOP:
+		/* invoke platform trigger to stop DMA even if pre_trigger failed */
```
Comment (plbossart) [inline sound/soc/sof/pcm.c:351]: I am just too confused here:

The comment on like 336 kinda hints that the DAI controls the DMA, but here it's the platform.

The commit message is equally convoluted. Everything seems to have triggers.

Lost....

Comment (ranj063) [reply sound/soc/sof/pcm.c:351]: @plbossart I should have done this a long time ago when we did the split for the HDA DAI ops. What this PR does is just mirror the same thing for the host DMA part.

Comment (plbossart) [reply sound/soc/sof/pcm.c:351]: it's even less clear to me what the comments referring to 'dai' mean if this is related to the HOST DMA management.

Comment (ranj063) [reply sound/soc/sof/pcm.c:351]: The DAI refers to the FE CPU DAI. I can change it to the Host DMA instead of DAI DMA in the comments if it makes it clearer.

File: sound/soc/sof/ipc3-pcm.c [hunk 4]
```diff
@@ -159,17 +159,52 @@ static int sof_ipc3_pcm_trigger(struct snd_soc_component *component,
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
 		stream.hdr.cmd |= SOF_IPC_STREAM_TRIG_PAUSE;
 		break;
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+		fallthrough;
```
Comment (plbossart) [inline sound/soc/sof/ipc3-pcm.c]: fallthrough not needed if there's nothing to do here.

File: sound/soc/sof/ipc4-pcm.c [hunk 5]
```diff
@@ -89,10 +89,24 @@ sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
...
+		switch (cmd) {
+		case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+		case SNDRV_PCM_TRIGGER_SUSPEND:
+		case SNDRV_PCM_TRIGGER_STOP:
+			/* Pause the pipeline if its started_count is 1 more than paused_count */
+			if (spipe->paused_count == (spipe->started_count - 1))
+				trigger_list->pipeline_ids[trigger_list->count++] =
+					pipe_widget->instance_id;
+			break;
+		case SNDRV_PCM_TRIGGER_START:
```
Comment (plbossart) [inline sound/soc/sof/ipc4-pcm.c:106]: not sure why TRIGGER_PAUSE_RELEASE is not handled here, is this intentional? If yes a comment would help.

File: sound/soc/sof/ipc3-pcm.c [hunk 6]
```diff
@@ -139,8 +139,8 @@ static int sof_ipc3_pcm_hw_params(struct snd_soc_component *component,
 	return ret;
 }
 
-static int sof_ipc3_pcm_trigger(struct snd_soc_component *component,
-				struct snd_pcm_substream *substream, int cmd)
+static int sof_ipc3_pcm_pre_trigger(struct snd_soc_component *component,
+				    struct snd_pcm_substream *substream, int cmd)
```
Comment (plbossart) [inline sound/soc/sof/ipc3-pcm.c:143]: commit message: I've seen all these explanations before, and we've already upstreamed code to fix things related to ordering already. It appears they were not good enough, so for the second pass I think it's required to explain what the shortcomings of the previous solution were, and why this one is better.

**Review by plbossart — CHANGES_REQUESTED** [review]
We'll need more background on why [PATCH 3/3] ASoC: SOF: pcm: Improve the pcm trigger sequence is not good enough and what problems were missed.

I also wonder if it's not better to revert and restart, than build on something that appears broken?

File: sound/soc/sof/ipc4-pcm.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c:106]: state can never be PAUSED when we're doing a release

**Comment by ranj063** [comment]
SOFCI TEST

File: sound/soc/sof/ipc4-pcm.c [hunk 7]
```diff
@@ -421,13 +423,42 @@ static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
...
 static int sof_ipc4_pcm_hw_free(struct snd_soc_component *component,
 				struct snd_pcm_substream *substream)
 {
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;
+
+	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);
+
+	/* nothing to trigger if the list is empty */
+	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:446]: This is a pcm_hw_free() callback, it is really odd to see things which 'trigger' at this phase.

File: sound/soc/sof/ipc4-pcm.c [hunk 8]
```diff
@@ -421,13 +423,42 @@ static int sof_ipc4_pcm_trigger(struct snd_soc_component *component,
...
+
+	/*
+	 * If use_chain_dma attribute is set we proceed to chained DMA
+	 * trigger function that handles the rest for the substream.
+	 */
+	if (sof_ipc4_is_chain_dma_in_use(pipeline_list))
+		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, SOF_IPC4_PIPE_RESET, 0);
+
 	/* command is not relevant with RESET, so just pass 0 */
 	return sof_ipc4_trigger_pipelines(component, substream, SOF_IPC4_PIPE_RESET, 0);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:457]: If you would introduce:
```c
static int sof_ipc4_change_pipelines_state(struct snd_soc_component *component,
					   struct snd_pcm_substream *substream,
					   int state, int cmd)
{
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;

	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);

	/* list is empty, nothing to do */
	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
		return 0;

	/* Use either ChainDMA or pipeline triggering function to set the new state */
	if (sof_ipc4_is_chain_dma_in_use(pipeline_list))
		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);

	return sof_ipc4_trigger_pipelines(component, substream, state, cmd);
}
```
then here you could call it:
```c
	/* command is not relevant with RESET, so just pass 0 */
	sof_ipc4_change_pipelines_state(component, substream, SOF_IPC4_PIPE_RESET, 0);
```
and use the same helper in the trigger function.
I think it would make the code cleaner and self explanatory in most places.

Comment (ujfalusi) [reply sound/soc/sof/ipc4-pcm.c:457]: and while at it, it might worth to change the `sof_ipc4_trigger_pipelines()` parameter list to `sdev, pipeline_list, state, cmd` also?

File: sound/soc/sof/ipc4-pcm.c [hunk 9]
```diff
@@ -89,10 +89,24 @@ sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
...
+			break;
+		case SNDRV_PCM_TRIGGER_START:
+			/* Pause the pipeline during start if it is starting for the first time */
+			if (!spipe->started_count)
+				trigger_list->pipeline_ids[trigger_list->count++] =
+					pipe_widget->instance_id;
+			break;
+		default:
+			break;
+		}
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:114]: Is this chunk related to #topic of the patch?

File: sound/soc/sof/ipc4-pcm.c [hunk 10]
```diff
@@ -463,6 +452,38 @@ static int sof_ipc4_pcm_hw_free(struct snd_soc_component *component,
 	return sof_ipc4_trigger_pipelines(component, substream, SOF_IPC4_PIPE_RESET, 0);
 }
 
+static int sof_ipc4_pcm_pre_trigger(struct snd_soc_component *component,
+				    struct snd_pcm_substream *substream, int cmd)
+{
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:462]: I would have expected to see in the file first the pre, then the post, not the other way around.

File: sound/soc/sof/ipc4-pcm.c [hunk 11]
```diff
@@ -463,6 +452,38 @@ static int sof_ipc4_pcm_hw_free(struct snd_soc_component *component,
...
+			return sof_ipc4_chain_dma_trigger(sdev, pipeline_list,
+							  SOF_IPC4_PIPE_PAUSED, 0);
+		return sof_ipc4_trigger_pipelines(component, substream, SOF_IPC4_PIPE_PAUSED, cmd);
+	default:
+		break;
+	}
+
+	return 0;
+}
+
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-pcm.c:491]: The helper I have proposed can be updated simplify the code to:
```c
static int sof_ipc4_change_pipelines_state(struct snd_soc_component *component,
					   struct snd_pcm_substream *substream,
					   int state, int cmd)
{
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
	struct snd_sof_pcm_stream_pipeline_list *pipeline_list;

	pipeline_list = sof_ipc4_get_pipeline_list(component, substream);

	/* list is empty, nothing to do */
	if (!pipeline_list || !pipeline_list->pipelines || !pipeline_list->count)
		return 0;

	/* Use either ChainDMA or pipeline triggering function to set the new state */
	if (sof_ipc4_is_chain_dma_in_use(pipeline_list)) {
		/* ChainDMA does not need to be placed in paused state on start */
		if (cmd == SNDRV_PCM_TRIGGER_START && state == SOF_IPC4_PIPE_PAUSED)
			return 0;

		return sof_ipc4_chain_dma_trigger(sdev, pipeline_list, state, cmd);
	}

	return sof_ipc4_trigger_pipelines(component, substream, state, cmd);
}
```

File: sound/soc/sof/sof-audio.h [hunk 12]
```diff
@@ -97,7 +97,8 @@ struct snd_sof_dai_config_data {
  * struct sof_ipc_pcm_ops - IPC-specific PCM ops
  * @hw_params: Function pointer for hw_params
  * @hw_free: Function pointer for hw_free
- * @trigger: Function pointer for trigger
+ * @post_trigger: Function pointer for host DMA post-trigger actions
+ * @pre_trigger: Function pointer for host DMA pre-trigger actions
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.h]: pre_trigger first, then post_trigger?

The comment should say in relation to **what** they are pre and post.

It is also odd that the host DMA is triggered via IPC on the DSP side, it is confusing to read it like this.
These are IPC specific  pre and post trigger callbacks around the platform trigger callback. If they start host DMA or not, it does not matter much, they do other things than that, right?

**Review by ujfalusi — COMMENTED** [review]
@ranj063, tests are looking great, but I think with some changes and clarifications this can be made better

File: sound/soc/sof/ipc4-pcm.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-pcm.c:106]: I would certainly hope that the only acceptable state is PAUSED when receiving a TRIGGER_PAUSE_RELEASE.

Not getting the explanation @ranj063?

File: sound/soc/sof/sof-audio.h (hunk 12 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.h]: I had the same thought on pre- and post- actions, they are missing a description of what they are anchored on. It looks like they are just two steps though?

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.h]: My guess is that it is a bit of both.
IPC3 have two sequences: IPC first, platform second or platform first, IPC second
IPC4 on the other hand have a twist in some trigger: IPC1 first, platform second, IPC2 second ?

In any case a clear description on what these pre/post triggers are relating to could help and in the IPC specific code a further comment can put the pieces into context.

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
replaced by https://github.com/thesofproject/linux/pull/4268
